spki/src/spki.rs

```diff
@@ -57,3 +68,34 @@ impl<'a> Message<'a> for SubjectPublicKeyInfo<'a> {
         f(&[&self.algorithm, &BitString::new(self.subject_public_key)?])
     }
 }
+
+#[cfg(feature = "fingerprint")]
+#[cfg_attr(docsrs, doc(cfg(feature = "fingerprint")))]
+#[derive(Debug)]
+pub enum FingerprintError {
+    DerEncodingError(Error),
+    Base64EncodingError(InvalidLengthError),
+}
+
+#[cfg(feature = "fingerprint")]
+#[cfg_attr(docsrs, doc(cfg(feature = "fingerprint")))]
+impl<'a> SubjectPublicKeyInfo<'a> {
+    /// Calculate the SHA-256 fingerprint of this SubjectPublicKeyInfo
+    pub fn fingerprint(
+        &self,
+        fingerprint: &'a mut [u8],
+    ) -> core::result::Result<&'a str, FingerprintError> {
```

Comment:
For a non-allocating API, I would suggest returning `digest::Output<Sha256>` which is effectively a byte array containing the raw digest.

Adding Base64 encoding support seems OK, but for that I'd probably just suggest adding an additional method like `fingerprint_base64` which returns a `String` and gating it on the `alloc` feature.

Reply:
Returning `digest::Output<Sha256>` was my initial implementation, however [RFC7469 §2.4](https://datatracker.ietf.org/doc/html/rfc7469#section-2.4) states:
> An SPKI Fingerprint is defined as the output of a known cryptographic
   hash algorithm whose input is the DER-encoded ASN.1 representation of
   the Subject Public Key Info (SPKI) of an X.509 certificate.  A Pin is
   defined as the combination of the known algorithm identifier and the
   SPKI Fingerprint computed using that algorithm.

>   The SPKI Fingerprint is encoded in base 64 for use in an HTTP header
   [RFC4648].

which is of course distinct from the X.509 AKI and SKI fields ([RFC5280 §4.2.1](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1)), which use the SHA-1 digest bytes directly, or the generally recognised concept of a "certificate thumbprint" which does not appear in RFC5280 at all but I've seen implemented widely as a SHA-1 hash and less widely as SHA-256.

So my reasoning here is to stick to the definition of a SPKI fingerprint and return Base64 only. I'm happy to additionally gate this on `alloc`, return a `String` and remove it from the non-allocating API entirely.

Reply:
I should also note that this only returns the fingerprint, not a full Pin including an `AlgorithmIdentifier`. 

Adding that may be a future enhancement

Reply:
All I'm seeing there re: Base64 is:

> The SPKI Fingerprint is encoded in base 64 for use in an HTTP header [RFC4648].

So I think it's fine to have a couple methods like:

- `fingerprint_bytes` (or maybe just call it `fingerprint`)
- `fingerprint_base64` (gated on `alloc`)

...and the latter is what you'd use as an HTTP header

Reply:
My read of that is that a SPKI fingerprint is encoded in Base64 as it is intended to be used in a HTTP header. 

Doesn't matter. It's not a lot of work to split out the functions. Watch this space.